File: LeetCode/129-sum-root-to-leaf-numbers.py
###############################################################################################
# 遍历一遍
###########
# 时间复杂度：O(n)
# 空间复杂度：O(logn)
###############################################################################################
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        res = 0
        def dfs(node, s):
            if not node.left and not node.right:
                nonlocal res
                res += int(s)
                return
            
            if node.left:
                dfs(node.left, s+f"{node.left.val}")
            if node.right:
                dfs(node.right, s+f"{node.right.val}")
        
        dfs(root, f"{root.val}")
        return res

File: LeetCode/129-sum-root-to-leaf-numbers.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// easy
////////////////
// Time Complexity: O(n)
// Space Complexity: O(logn)
///////////////////////////////////////////////////////////////////////////////////////////////
#define ll long long
class Solution {
public:
    ll ans = 0;
    void dfs(TreeNode *root, ll tmp){
        if(root->left== NULL && root->right==NULL){
            tmp = tmp*10 + root->val;
            ans += tmp;
            return;
        }
        tmp = tmp*10 + root->val;
        if(root->left) dfs(root->left, tmp);
        if(root->right) dfs(root->right, tmp);
    }
    int sumNumbers(TreeNode* root) {
        dfs(root,0);
        return int(ans);
    }
};